src/Input/InputParamTrait.php

```diff
@@ -0,0 +1,138 @@
+<?php
+
+/**
+ * @see       https://github.com/laminas/laminas-cli for the canonical source repository
+ * @copyright https://github.com/laminas/laminas-cli/blob/master/COPYRIGHT.md
+ * @license   https://github.com/laminas/laminas-cli/blob/master/LICENSE.md New BSD License
+ */
+
+declare(strict_types=1);
+
+namespace Laminas\Cli\Input;
+
+use InvalidArgumentException;
+use Symfony\Component\Console\Input\InputOption;
+
+use function sprintf;
+
+/**
+ * Provide the majority of methods needed to implement InputParamInterface.
+ *
+ * This trait provides definitions for all but the following methods of the
+ * InputParamInterface:
+ *
+ * - getOptionMode()
+ * - getQuestion()
+ *
+ * Additionally, it defines the `$name` property, allowing implementations to
+ * set it in their constructors without needing to define the property
+ * themselves.
+ */
+trait InputParamTrait
```

Comment:
I am looking on it again, and I am thinking now, that maybe it will be better if we change it to abstract class and extend [`InputOption`](https://github.com/symfony/console/blob/master/Input/InputOption.php).

In general we are duplicating many functions from `InputOptions`, so I think it would make sense to extend it, and change + extends to accomplish our goal for InputParams.

For example, please note default values - we do not have any checks here, but there are some checks in InputOption. If we provide any value it will fail at some point, as we are adding the option to the command anyway. It would be nice if it fails on creation of the InputParam, not on adding to param to the command.

Sorry, I haven't noted it before.

Reply:
Don't apologize! This is what review is for, and the amount of review I've had the past week has continually made this approach better!

I think this is a great idea; I'll start work on it today.

Reply:
I've been working on this off and on during meetings today, and I've come to the conclusion this isn't going to work.

`InputOption` defines each of the description, mode, and shortcut as `private`, and does not provide setters for any of those values. **This makes it impossible to read or write to those values from an extending class.** As a result, we are left with the following options:

- We inherit our constructor from `InputOption`. This basically goes against the entire design I've put forward for the InputParams, whereby they act as builders, and the constructor is only used for **required data**, such as the name, the haystack for choices, the path type for paths, etc. I think this would be step back in usability, and make it far harder for us to have consistent constructors.

- We add something like an `initializeOption()` method that the `AbstractParamAwareCommand` calls during `addParam()` to populate the option, and have it call on `parent::__construct()` with related values. The problem here is that we can no longer force the default value to be `null`, which then leads to issues when determining whether to prompt or not. Additionally, it makes finding _when_ an invalid value was introduced harder and/or leads to code duplication and/or puts us in a situation where our signatures might conflict with upstream signatures. For instance, if I provide a `setShortcut()` method, I need to validate and normalize the value in the setter, as delaying it to `initializeOption()` will make it harder for users to debug (that call will not lead them back to the code where they created the parameter instance). Since `InputOption` does that validation and normalization _in its constructor_, I end up duplicating work. On top of that, there is method overlap with methods like `getDescription()` where I define a return value, but `InputOption` does not; while this works _now_, we cannot guarantee it will work _in the future_.

With regards to `setDefault()`, the checks you refer to are all based on _preceding constructor arguments_. Essentially, it is performing checks based on the _mode_; this leads to an order of operations issue, whereby we need to enforce setting the mode before setting the default value. This gets even more problematic because _we want the input option to behave as if it does not have a default value_! This is how we are able to detect in non-interactive mode that a value is not present, and thus return the default. Even if I work around that by supplying a different method:

- `setDefault()` is defined on the public API of `InputOption`, so users might inadvertently call it instead of a method we define.
- we still run into the order of operations issue in terms of validating the default value in terms of the mode selected; it has to happen either at option initialization, or when we call `getDefault()`, both of which are removed from creation of the param in the first place.

My point is: the design of the `InputOption` class makes it next-to-impossible to extend in meaningful ways, and it actively gets in the way of what we want to do.

Having a separate type entirely prevents most of these issues, because we seed an `InputOption` from the param. It's not perfect, though, because:

- We never validate the default value (since it's not injected into the `InputOption`, either)
- We limit what is accepted by parameters for shortcut values (though we could expand this to mirror the `InputOption`)

There is one last possibility I can think of: I create a builder class that provides a method to return an instance, and that instance will be an `InputOption` implementing an `InputParamInterface` where only the methods `getQuestion()` and `getParamDefault()` are defined:

```php
$param = (new StringParam('test'))
    ->setDescription('Prompting for a string')
    ->setShortcut('s')
    ->setDefault('default string')
    ->setPattern('/^[a-z0-9_ -]+$/')
    ->createParam();
```

This would allow us to re-use the validation in the `InputOption` constructor, and, generally speaking, help users narrow down where in the code the problem occurred (as long as they call `createParam()` in the same code location as where they add option/parameter values). The builder would be a true builder at that point (instead of also containing behavior), which would likely satisfy some concerns @boesing raised last week.

Flip side: it's a big change, and my time is more limited this week, so it puts us back a bit on getting this feature released.

I'm personally fine with the way it _currently_ works, but I also see the value of keeping the API close to how `InputOption` works. If the builder pattern I propose above looks reasonable, I'll start work on it this week.

Reply:
@weierophinney Thanks for that investigation. I see it is problematic extending InputOption and I completely forgot about that we are not setting there default value.

I do not like "builders" as it look like mock builder from PHPUnit ;-)

I think, we should keep it as it is, just:
- extend shortcuts to mirror InputOptions (it would be helpful if we have a wrappers)
- change trait to AbstractInputParam class (maybe leave only questions as traits?), as I think it is more common way to extend some abstract. In general I do not like traits too much, as these could cause many issues - we cannot provide that trait can be used in specific interface etc... 

And - as we are still planning 0.1.0 release first, that's it. We still always can make bigger BC changes for the next 0.x release if it will be needed, and I think we should notice it when more people start using it, and when we start writing our own commands in other components.